### PR TITLE
Fixed failing Date/time and math functions tests with Oracle DB.

### PR DESCRIPTION
**Pull request by Tomas-Kraus**
- fixes issue #1539
- fixes issue #1540
- fixes issue #1541

All LRG tests still wont pass with Oracle DB but all known problems with date/time and math functions should be fixed now.

### CODE REVIEW

**Comment by Tomas-Kraus** [comment]
@jgrassel Please verify that problem is resolved in your environment too. I used Oracle 21c to run the tests locally.

File: foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/databaseaccess/DatabasePlatform.java [hunk 1]
```diff
@@ -1450,6 +1450,25 @@ public Object getObjectFromResultSet(ResultSet resultSet, int columnNumber, int
...
+     *
+     * @param resultSet JDBC {@code ResultSet} with query result
+     * @param field database field mapping
+     * @param type database type from {@link java.sql.Types} or provider specific type
+     * @param columnNumber number of column to fetch
+     * @param session current database session
+     * @param unOptimized original unOptimized value to be returned if no optimized conversion is possible
+     * @return new ResultSet column value or original unOptimized value if no conversion was done.
+     */
+    public Object getObjectThroughOptimizedDataConversion(
```
Comment (lukasj) [inline foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/databaseaccess/DatabasePlatform.java:1466]: Addition of this method looks wrong and should not be needed. The mapping between Java and DB types is defined in  `platform.buildFieldTypes()` method and `getObjectThroughOptimizedDataConversion` is supposed to work with what that method returns

Comment (Tomas-Kraus) [reply foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/databaseaccess/DatabasePlatform.java:1466]: It depends. `buildFieldTypes()` defines mapping from Java to SQL types for DB table creation. `getObject...` code handles mapping between `ResultSet` column and `DatabaseField` of target Java value. In this case its 1:N where SQL side is for example `TIME` and `DatabaseField` may be any of Local/Offset/sql variant of such a field.
I can move this conversion to `DatabaseAccessor` (where most of such a code already is), but direct conversion from returned Oracle `TIMESTAMPTZ`/`TIMESTAMPLTZ` seems better for me.

Comment (Tomas-Kraus) [reply foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/databaseaccess/DatabasePlatform.java:1466]: So if you think that adding this mehod has less benefits than problems, I'll solve it in `DatabaseAccessor`. Just let me know.

Comment (Tomas-Kraus) [reply foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/databaseaccess/DatabasePlatform.java:1466]: Removal of this code caused:
```
[ERROR] Errors: 
[ERROR]   TestDateTimeFunctions.testCriteriaQuerySelectLocalDateTime:456 ClassCast class...
[ERROR]   TestDateTimeFunctions.testCriteriaQuerySelectLocalTime:402 ClassCast class org...
[ERROR]   TestOracleLOBLocatorFeature.testOracleLOBLocatorWithEmptyClob:136 » Rollback E...
[INFO] 
[ERROR] Tests run: 659, Failures: 0, Errors: 3, Skipped: 12

[ERROR] org.eclipse.persistence.jpa.test.criteria.TestDateTimeFunctions.testCriteriaQuerySelectLocalTime  Time elapsed: 0.019 s  <<< ERROR!
java.lang.ClassCastException: class org.eclipse.persistence.internal.platform.database.oracle.TIMESTAMPTZWrapper cannot be cast to class java.time.LocalTime (org.eclipse.persistence.internal.platform.database.oracle.TIMESTAMPTZWrapper is in unnamed module of loader 'app'; java.time.LocalTime is in module java.base of loader 'bootstrap')
	at org.eclipse.persistence.jpa.test.criteria.TestDateTimeFunctions.testCriteriaQuerySelectLocalTime(TestDateTimeFunctions.java:402)

[ERROR] org.eclipse.persistence.jpa.test.criteria.TestDateTimeFunctions.testCriteriaQuerySelectLocalDateTime  Time elapsed: 0.014 s  <<< ERROR!
java.lang.ClassCastException: class org.eclipse.persistence.internal.platform.database.oracle.TIMESTAMPTZWrapper cannot be cast to class java.time.LocalDateTime (org.eclipse.persistence.internal.platform.database.oracle.TIMESTAMPTZWrapper is in unnamed module of loader 'app'; java.time.LocalDateTime is in module java.base of loader 'bootstrap')
	at org.eclipse.persistence.jpa.test.criteria.TestDateTimeFunctions.testCriteriaQuerySelectLocalDateTime(TestDateTimeFunctions.java:456)
```
So `TIMESTAMPTZWrapper` is being returned without conversion but target variable is `java.time.LocalDateTime`.

Comment (Tomas-Kraus) [reply foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/databaseaccess/DatabasePlatform.java:1466]: This means, that I have to handle Oracle specific type anyway. One option is keep it as it it now, the other one is to add Oracle specific SQL types:
```
    public static final int TIMESTAMPTZ  = -101;
    public static final int TIMESTAMPLTZ = -102;
```
to be handled together with 
```
     public static final int DATE            =  91;
     public static final int TIME            =  92;
     public static final int TIMESTAMP       =  93;
```
in `DatabaseAccessor`, but think that Oracle specific thnings shall not be there.

Comment (Tomas-Kraus) [reply foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/databaseaccess/DatabasePlatform.java]: Replaced this with conversion in `Oracle9Platform#convertObject(Object, Class<T>)`

**Comment by rfelcman** [comment]
LGTM (JPA WDF test failures are known non-related issue in EL PR verifier).

**Comment by Tomas-Kraus** [comment]
- fixes issue #1547
- fixes issue #1548
- fixes issue #1549
- fixes issue #1550

**Pull request merged**